website/project/model.py

```diff
@@ -2530,3 +2576,45 @@ def to_json(self):
                       for x in self.nodes if not x.is_deleted],
             "anonymous": self.anonymous
         }
+
+
+class Retraction(StoredObject):
+    """Retraction object for public registrations."""
+
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    justification = fields.StringField(default=None, validate=MaxLengthValidator(2048))
+    initiation_date = fields.DateTimeField()
+    initiated_by = fields.ForeignField('user', backref='retracted_by')
+    approval_state = fields.DictionaryField()
+    state = fields.StringField(default='pending')
+
+    @property
+    def is_retracted(self):
+        return self.state == 'retracted'
+
+    @property
+    def pending_retraction(self):
+        return self.state == 'pending'
+
+    def disapprove_retraction(self, user, token):
+        """Cancels retraction if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['disapproval_token'] != token:
+                raise exceptions.InvalidRetractionDisapprovalToken
+            self.state = 'cancelled'
+        except KeyError:
+            raise ValidationValueError('User must be an admin to disapprove retraction of a registration.')
+
+    def approve_retraction(self, user, token):
+        """Add user to approval list if user is admin and token verifies."""
+        try:
+            if self.approval_state[user._id]['approval_token'] != token:
+                raise exceptions.InvalidRetractionApprovalToken
+            self.approval_state[user._id]['has_approved'] = True
+            num_of_approvals = self.approval_state['num_of_approvals']
+            self.approval_state['num_of_approvals'] = num_of_approvals + 1
```

Comment:
Is it necessary to maintain `num_of_approvals` when it can be computed from the dictionary values?

Something like: 

```python
# caveat: untested
sum(val['has_approved'] for val in self.approval_state.values())  
```